### ISSUE TITLE
over/underflow in zeta for complex argumetns

### ISSUE THREAD

**Issue by stevengj** [JuliaMath/SpecialFunctions.jl#128]
As reported [on discourse](https://discourse.julialang.org/t/possible-issue-with-zeta-function/16802) by @ArchieCall
```jl
julia> zeta(.4 + 453.0im)
NaN + NaN*im
```
The error seems to be triggered by this line:
https://github.com/JuliaMath/SpecialFunctions.jl/blob/803c96955356995d4d063a87e876961033502cbd/src/gamma.jl#L412

In particular, the problem is the `gamma(1 - s) * sinpi(s*0.5)` expression in the reflection formula.  If `s` has a large imaginary part, this is the product of an exponentially small with an exponentially large term.   Eventually, `gamma` underflows and `sinpi` overflows, giving `0*Inf = NaN`.

The solution would be to refactor the exponential terms and cancel them analytically.

**Comment by stevengj** [comment]
Looks like it should suffice to replace the coefficient of `zeta(1 - s)` with:
```jl
absim = abs(imag(s))
if absim > 12 # amplitude of sinpi(s/2) ≈ exp(imag(s)*π/2)
    lg = lgamma(1 - s)
    ln2pi = 1.83787706640934548356 # log(2pi) to double precision
    rehalf = real(s)*0.5
    return exp(lg + absim*(pi/2) + s*ln2pi) *
           (Complex(sinpi(rehalf), copysign(cospi(rehalf), imag(s))) * (0.5/π))
else
    return gamma(1 - s) * sinpi(s*0.5) * (2π)^s / π
end
```

**Comment by ArchieCall** [comment]
Steve,  Thanks for getting back on this issue.  I am not an expert on Git so I just commented out line 412 on my local machine and pasted in your code from above.
When I went back into Julia, it recompiled SpecialFunctions, so I think the code was in the proper location

It does not blow off now with higher values of the imaginary value.  However, the output values do not agree with WolframAlpha.
...Archie
